### PR TITLE
Adding PageActionDescriptorProvider

### PR DESCRIPTION
**Pull request by pranavkm**

### CODE REVIEW

**Comment by pranavkm** [comment]
cc @rynowak

File: src/Microsoft.AspNetCore.Mvc.RazorPages/ApplicationModels/IPageModelConvention.cs [hunk 1]
```diff
@@ -0,0 +1,21 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+namespace Microsoft.AspNetCore.Mvc.RazorPages.ApplicationModels
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.RazorPages/ApplicationModels/IPageModelConvention.cs]: Drop the `.RazorPages` IMO

File: src/Microsoft.AspNetCore.Mvc.RazorPages/Infrastructure/DefaultRazorProject.cs [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using Microsoft.Extensions.FileProviders;
+
+namespace Microsoft.AspNetCore.Mvc.RazorPages.Infrastructure
+{
+    public class DefaultRazorProject : RazorProject
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.RazorPages/Infrastructure/DefaultRazorProject.cs]: Should we keep this? Let's discuss. This was part of the prototype that I wasn't 100% confident in.

File: src/Microsoft.AspNetCore.Mvc.RazorPages/Infrastructure/PageActionDescriptorProvider.cs [hunk 3]
```diff
@@ -0,0 +1,191 @@
...
+
+            foreach (var selector in model.Selectors)
+            {
+                actions.Add(new PageActionDescriptor()
+                {
+                    AttributeRouteInfo = new AttributeRouteInfo()
+                    {
+                        Name = selector.AttributeRouteModel.Name,
+                        Order = selector.AttributeRouteModel.Order ?? 0,
+                        Template = selector.AttributeRouteModel.Template,
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.RazorPages/Infrastructure/PageActionDescriptorProvider.cs]: This needs error handling.

File: src/Microsoft.AspNetCore.Mvc.RazorPages/Infrastructure/PageActionDescriptorProvider.cs [hunk 4]
```diff
@@ -0,0 +1,191 @@
...
+            }
+            else if (prefix == string.Empty)
+            {
+                return template;
+            }
+            else
+            {
+                return prefix + "/" + template;
+            }
+        }
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.RazorPages/Infrastructure/PageActionDescriptorProvider.cs]: I think we already have a lot of this code somewhere. Look at the `AttributeRouting` class.

File: src/Microsoft.AspNetCore.Mvc.RazorPages/Infrastructure/PageActionDescriptorProvider.cs [hunk 5]
```diff
@@ -0,0 +1,191 @@
...
+
+        private struct RouteTemplates
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.RazorPages/Infrastructure/PageActionDescriptorProvider.cs]: This is kinda gross, and isn't close to what's really being done here. My suggestion would be to pass in the `PageModel` and add `SelectorModels` to it.

File: src/Microsoft.AspNetCore.Mvc.RazorPages/Infrastructure/PageDirectiveFeature.cs [hunk 6]
```diff
@@ -0,0 +1,32 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.IO;
+
+namespace Microsoft.AspNetCore.Mvc.RazorPages.Infrastructure
+{
+    public static class PageDirectiveFeature
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.RazorPages/Infrastructure/PageDirectiveFeature.cs]: This is OK as a temporary hack

Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.RazorPages/Infrastructure/PageDirectiveFeature.cs]: Please file an item to get rid of it though 😆

Comment (pranavkm) [inline src/Microsoft.AspNetCore.Mvc.RazorPages/Infrastructure/PageDirectiveFeature.cs]: https://github.com/aspnet/Mvc/issues/5475

**Comment by pranavkm** [comment]
🆙 📅

File: src/Microsoft.AspNetCore.Mvc.Core/ApplicationModels/AttributeRouteModel.cs [hunk 7]
```diff
@@ -99,6 +99,33 @@ public bool IsAbsoluteTemplate
             };
         }
 
+        /// <summary>
+        /// Combines the prefix and route template for an attribute route.
+        /// </summary>
+        /// <param name="prefix">The prefix.</param>
+        /// <param name="template">The route template.</param>
+        /// <returns>The combined pattern.</returns>
+        public static string CombineTemplates(string prefix, string template)
```
Comment (javiercn) [inline src/Microsoft.AspNetCore.Mvc.Core/ApplicationModels/AttributeRouteModel.cs]: Why are we making this logic public? Doesn't seem related to the change

File: src/Microsoft.AspNetCore.Mvc.RazorPages/Infrastructure/PageDirectiveFeature.cs [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+    {
+        public static bool TryGetRouteTemplate(RazorProjectItem projectItem, out string template)
+        {
+            const string PageDirective = "@page";
```
Comment (javiercn) [inline src/Microsoft.AspNetCore.Mvc.RazorPages/Infrastructure/PageDirectiveFeature.cs]: Does this need to be at the very beginning of a razor page? What if my page has a whitespace character at the beginning? like `____@page` (replace _ for whitespaces)

File: src/Microsoft.AspNetCore.Mvc.RazorPages/Infrastructure/PageDirectiveFeature.cs [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+
+            string content;
+            using (var streamReader = new StreamReader(projectItem.Read()))
+            {
+                content = streamReader.ReadToEnd();
+            }
+
+            if (content.StartsWith(PageDirective, StringComparison.Ordinal))
```
Comment (javiercn) [inline src/Microsoft.AspNetCore.Mvc.RazorPages/Infrastructure/PageDirectiveFeature.cs]: The answer is most likely no, given that you are the expert you definitely known better. This is in essence a mini/super incomplete parser for razor, would it be possible to use razor's parser for this taks? (Or lexer if it has one). I'm concerned for changes in behavior in this area would need to be replicated here, I also can see users putting `@page` after things like @using or @model, and as with any code involving parsing, all the small details about it (whitespace handling, etc) that are already baked in the parser and we should reuse if possible

File: src/Microsoft.AspNetCore.Mvc.RazorPages/Infrastructure/PageActionDescriptorProvider.cs [hunk 10]
```diff
@@ -0,0 +1,125 @@
...
+            }
+        }
+
+        private static SelectorModel CreateSelectorModel(string prefix, string template)
+        {
+            return new SelectorModel
+            {
+                AttributeRouteModel = new AttributeRouteModel
+                {
+                    Template = AttributeRouteModel.CombineTemplates(prefix, template),
```
Comment (javiercn) [inline src/Microsoft.AspNetCore.Mvc.RazorPages/Infrastructure/PageActionDescriptorProvider.cs]: I think you can drastically simplify this. As far as I understand, prefix is never going to be null, and template does not support override patterns so you can just do `template != null ? (prefix.EndsWith("/") ? prefix + template : $"{prefix}/{template}") : prefix;`

I think we shouldn't be using `AttributeRouteModel.CombineTemplates` here. It is basically for describing how to combine routes when created through attributes on the controller and action. Reusing it here feels wrong.

File: src/Microsoft.AspNetCore.Mvc.Core/ApplicationModels/AttributeRouteModel.cs (hunk 7 above)
Comment (javiercn) [inline src/Microsoft.AspNetCore.Mvc.Core/ApplicationModels/AttributeRouteModel.cs]: See comment below, we shouldn't reuse this for razor pages. No need to add more public API surface for this

File: src/Microsoft.AspNetCore.Mvc.Core/ApplicationModels/AttributeRouteModel.cs [hunk 11]
```diff
@@ -99,6 +99,33 @@ public bool IsAbsoluteTemplate
...
+
+        /// <summary>
+        /// Determines if a template pattern can be used to override a prefix.
+        /// </summary>
+        /// <param name="template">The template.</param>
+        /// <returns><c>true</c> if this is an overriding template, <c>false</c> otherwise.</returns>
+        /// <remarks>
+        /// Route templates starting with "~/" or "/" can be used to override the prefix.
+        /// </remarks>
+        public static bool IsOverridePattern(string template)
```
Comment (javiercn) [inline src/Microsoft.AspNetCore.Mvc.Core/ApplicationModels/AttributeRouteModel.cs]: Similar reason as above, see comment below. This is tied to Attribute routing and we shouldn't reuse it for razor pages.

File: src/Microsoft.AspNetCore.Mvc.RazorPages/PageActionDescriptor.cs [hunk 12]
```diff
@@ -0,0 +1,14 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using Microsoft.AspNetCore.Mvc.Abstractions;
+
+namespace Microsoft.AspNetCore.Mvc.RazorPages.Infrastructure
+{
+    public class PageActionDescriptor : ActionDescriptor
```
Comment (javiercn) [inline src/Microsoft.AspNetCore.Mvc.RazorPages/PageActionDescriptor.cs]: nit: It would be nice for this to have a friendly display name like ControllerActionDescriptor has

File: src/Microsoft.AspNetCore.Mvc.Core/ApplicationModels/AttributeRouteModel.cs (hunk 7 above)
Comment (pranavkm) [inline src/Microsoft.AspNetCore.Mvc.Core/ApplicationModels/AttributeRouteModel.cs]: @rynowak, how do you feel about these APIs? I'm not particularly vested in this, but this was my interpretation of not duplicating code per https://github.com/aspnet/Mvc/pull/5450#discussion_r85363884

**Comment by rynowak** [comment]
This PR is crazy for some reason. Regarding the discussion about these APIs, pages routing **is** attribute routing. I don't see the conflict.

**Comment by pranavkm** [comment]
Sorry, I had to rebase on dev because of the 1.1.0 packages that got pushed to nuget.org yesterday. I'll try resending the PR.
